telegram/ext/_application.py

```diff
@@ -0,0 +1,1224 @@
+                break
+
+            # Dispatch any error.
+            except Exception as exc:
+                if await self.dispatch_error(update, exc):
+                    _logger.debug('Error handler stopped further handlers.')
+                    break
+
+        if any_blocking:
+            # Only need to mark the update for persistence if there was at least one
+            # blocking handler - the non-blocking handlers mark the update again when finished
+            self._mark_for_persistence_update(update=update)
+
+    def add_handler(self, handler: Handler[Any, CCT], group: int = DEFAULT_GROUP) -> None:
+        """Register a handler.
+
+        TL;DR: Order and priority counts. 0 or 1 handlers per group will be used. End handling of
+        update with :class:`telegram.ext.ApplicationHandlerStop`.
+
+        A handler must be an instance of a subclass of :class:`telegram.ext.Handler`. All handlers
+        are organized in groups with a numeric value. The default group is 0. All groups will be
+        evaluated for handling an update, but only 0 or 1 handler per group will be used. If
+        :class:`telegram.ext.ApplicationHandlerStop` is raised from one of the handlers, no further
+        handlers (regardless of the group) will be called.
+
+        The priority/order of handlers is determined as follows:
+
+          * Priority of the group (lower group number == higher priority)
+          * The first handler in a group which should handle an update (see
+            :attr:`telegram.ext.Handler.check_update`) will be used. Other handlers from the
+            group will not be used. The order in which handlers were added to the group defines the
+            priority.
+
+        Args:
+            handler (:class:`telegram.ext.Handler`): A Handler instance.
+            group (:obj:`int`, optional): The group identifier. Default is 0.
+
+        """
+        # Unfortunately due to circular imports this has to be here
+        # pylint: disable=import-outside-toplevel
+        from telegram.ext._conversationhandler import ConversationHandler
+
+        if not isinstance(handler, Handler):
+            raise TypeError(f'handler is not an instance of {Handler.__name__}')
+        if not isinstance(group, int):
+            raise TypeError('group is not int')
+        if isinstance(handler, ConversationHandler) and handler.persistent and handler.name:
+            if not self.persistence:
+                raise ValueError(
+                    f"ConversationHandler {handler.name} "
+                    f"can not be persistent if application has no persistence"
+                )
+            if self._initialized:
+                warn(
+                    'A persistent `ConversationHandler` was passed to `add_handler`, '
+                    'after `Application.initialize` was called. Conversation states will not be '
+                    'loaded from persistence!',
```

Comment:
The problem with that is that we'd have to make `add_handler` async, since `BP.get_conversations` is async now. This doesn't seem desirable to me. What we could do is `self.create_task(conv._init_pers(self))` and issue a warning that with this it might happen that an update comes in before the persistence is loaded. What do you think?

I went with this approach for now, because the culprits here are rather subtle and adding persistent conversation handlers at runtime seemed a rare enough use case to me.

- btw, I realize that in `CH._init_pers`, we should use the `_conversations_lock`

Reply:
> What we could do is `self.create_task(conv._init_pers(self))` and issue a warning that with this it might happen that an update comes in before the persistence is loaded. What do you think?

Yes, sounds reasonable. 

> adding persistent conversation handlers at runtime seemed a rare enough use case to me.

quite a few people do add/remove handlers at runtime for some reason